### ISSUE TITLE
Rule proposal: Ban method override type narrowing

### ISSUE THREAD

**Issue by Woodz** [typescript-eslint/typescript-eslint#8964]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Validate that overriden method signatures match the base signature (i.e. no narrowing of types)

### Fail Cases

```TypeScript
class A {
  f(p: string | number) {
    console.log('A', p);
  }
}

class B extends A {
  f(p: string) {
    console.log('B', p, p.charCodeAt(0));
  }
}

const arrayOfA: Array<A> = [new A(), new B()];
for (const item of arrayOfA) {
  item.f(123);
}
```

### Pass Cases

```TypeScript
class A {
  f(p: string | number) {
    console.log('A', p);
  }
}

class B extends A {
  f(p: string | number) { <-- Needs to match base class method signature
    console.log('B', p, p.charCodeAt(0));
  }
}

const arrayOfA: Array<A> = [new A(), new B()];
for (const item of arrayOfA) {
  item.f(123);
}
```

### Additional Info

TS issue was closed as "working as intended" but this is very dangerous, since this breaks some type safety, meaning that this introduces a risk of runtime errors

**Comment by kirkwaiblinger** [comment]
Huh. Without really having thought about this, my initial impression is that I would have thought that TS would allow subclasses to accept _wider_ parameters and return _narrower_ return types. But if I understand correctly it allows only narrower types both for parameters and return values, is that right?

**Comment by Woodz** [comment]
> Huh. Without really having thought about this, my initial impression is that I would have thought that TS would allow subclasses to accept _wider_ parameters and return _narrower_ return types. But if I understand correctly it allows only narrower types both for parameters and return values, is that right?

Currently TS supports both wider and narrower parameter types. I don't think either should be allowed

**Comment by Josh-Cena** [comment]
The rule should not be solely confined to inheritance. IMO, it should also detect general assignability, but not sure if that's performant. Just suggesting the rule to at least be generically named.

**Comment by kirkwaiblinger** [comment]
On the naming, noting that the best technical term is probably "heritage methods" but we chose to use "inherited methods" as the user-facing terminology in https://github.com/typescript-eslint/typescript-eslint/pull/8765#discussion_r1623457327. Will probably want to be consistent here.

Also - v8 is now live, so we can start using `checker.isTypeAssignableTo` for this if it makes sense to do so.

**Comment by ronami** [comment]
Looking a bit into this, I think it's closely related (possibly even a duplicate) to #277.

In my opinion, the following examples should all be reported and belong under a single rule:

```ts
// --- Inheritance ---

export class A {
  f(_p: number | string): void {}
}

export class B extends A {
  // Property 'f' in type 'B' is not assignable to the same property in base type 'A'.
  f(_p: string): void {}
}

// --- Interfaces ---

export interface C {
  f(_p: number | string): void;
}

export interface D extends C {
  // Property 'f' in type 'D' is not assignable to the same property in base type 'C'.
  f(_p: string): void;
}

// --- Interface implementation ---

export interface E {
  f(_p: number | string): void;
}

export class F implements E {
  // Property 'f' in type 'F' is not assignable to the same property in base type 'E'.
  f(_p: string): void {}
}

// --- Assignment ---

class G {
  f(_p: number | string): void {}
}

const e = new G();
// Property 'f' is not assignable to the same property in base type 'G'.
e.f = (_p: number): void => {};
```

Based on the above, I'm personally not sure if there's a reason to have both issues open.

**Comment by JoshuaKGoldberg** [comment]
👍 that's a good point. I'd be in favor of combining these two issues into one. It might make sense to rename #277 to be more specific?

**Comment by ronami** [comment]
Yes, I think so too 👍 

I'm having a hard time thinking of a fitting name. I like @kirkwaiblinger's suggestion to use "heritage methods" the most (in https://github.com/typescript-eslint/typescript-eslint/issues/8964#issuecomment-2264239677).

**Comment by kirkwaiblinger** [comment]
> I'm having a hard time thinking of a fitting name. I like @kirkwaiblinger's suggestion to use "heritage methods" the most (in [#8964 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/8964#issuecomment-2264239677)).

I actually meant to suggest we _not_ use "heritage methods" in a user-facing context if we can avoid it. It's a technical term that makes sense to us but feels super daunting to me as a user. Hence the decision to go with "inherited methods" in https://github.com/typescript-eslint/typescript-eslint/pull/8765#discussion_r1623457327 (after a rather lengthy discussion).

Let's continue conversation on #277
